### PR TITLE
Support collecting http parameters

### LINKED ISSUE #324: Add option to start/stop collecting http get/post parameter
cc @ascrutae @pengys5 

I want to add this feature, after we have config downstream mechanism.

### PR DESCRIPTION
**Pull request by kezhenxu94**
Plugins of Tomcat, Spring MVC, Armeria are supported for now

Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

This pull request provides another config item to enable/disable collecting http parameters of a request.

It uses a general name `plugin.http.collect_http_params` but doesn't guarantee that all http plugins should be under its control, unless the plugin itself takes this config item into consideration, such as Tomcat, Spring MVC, and Armeria plugin.

This pull request also partially close #324 but I would suggest open another issue to track the downstreaming config feature, and I believe we need more discussion on that, @wu-sheng

### CODE REVIEW

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/util/CollectionUtil.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * Reinvent the wheels because importing third-party libs just for some methods is not worthwhile in agent side
+ *
+ * @author kezhenxu94
+ * @since 7.0.0
+ */
+public final class CollectionUtil {
+    public static String toString(final Map<String, String[]> map) {
+        return map
+            .entrySet()
+            .parallelStream()
```
Comment (dmsolr) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/util/CollectionUtil.java]: As far as I know, `parallelStream()` could cause high latency on high requests.

**Comment by kezhenxu94** [comment]
> Why change Tomcat plugin rather than changing SpringMVC plugin? It should be more adaptive?

Is it possible to assume that all http services are based on SpringMVC? Armeria is an apparent exception, and that's why we have Tomcat plugin and Spring MVC plugin now

**Comment by wu-sheng** [comment]
I am not sure whether Tomcat plugin will face the parameter encoding issue.

**Comment by kezhenxu94** [comment]
> I am not sure whether Tomcat plugin will face the parameter encoding issue.

Yes sadly, I've suffered from that, and that's why the codes to get parameter map in Tomcat plugin is more complicated than those in Spring MVC's, at least I've tested and the plugin test codes show that.

**Comment by kezhenxu94** [comment]
And is running on my test environment now

File: docs/en/setup/service-agent/java-agent/README.md [hunk 2]
```diff
@@ -118,6 +118,8 @@ property key | Description | Default |
 `plugin.opgroup.*`|Support operation name customize group rules in different plugins. Read [Group rule supported plugins](op_name_group_rule.md)|Not set|
 `plugin.springtransaction.simplify_transaction_definition_name`|If true, the transaction definition name will be simplified.|false|
 `plugin.jdkthreading.threading_class_prefixes` | Threading classes (`java.lang.Runnable` and `java.util.concurrent.Callable`) and their subclasses, including anonymous inner classes whose name match any one of the `THREADING_CLASS_PREFIXES` (splitted by `,`) will be instrumented, make sure to only specify as narrow prefixes as what you're expecting to instrument, (`java.` and `javax.` will be ignored due to safety issues) | Not set |
+`plugin.http.collect_http_params`| This config item controls that whether the plugins related to HTTP should collect the parameters of the request. The name `plugin.http.collect_http_params` is rather general, but it doesn't guarantee that all http plugins should be under its control, unless the plugin itself takes this config item into consideration, such as Tomcat, Spring MVC, and Armeria plugin. | `false` |
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/README.md]: Do we need to indicate this should work for UTF-8 encoding project? I think some vendors are using GBK in telecom, bank, etic

Comment (kezhenxu94) [reply docs/en/setup/service-agent/java-agent/README.md]: > Do we need to indicate this should work for UTF-8 encoding project? I think some vendors are using GBK in telecom, bank, etic

I've modified the case to specify the encoding as GBK, and it works, still

File: test/plugin/scenarios/spring-3.1.x-scenario/src/main/webapp/WEB-INF/web.xml [hunk 3]
```diff
@@ -23,6 +23,20 @@
...
+            <param-name>encoding</param-name>
+            <param-value>GBK</param-value>
+        </init-param>
+    </filter>
+
+    <filter-mapping>
+        <filter-name>setCharacterEncodingFilter</filter-name>
+        <url-pattern>/*</url-pattern>
+    </filter-mapping>
+
```
Comment (kezhenxu94) [inline test/plugin/scenarios/spring-3.1.x-scenario/src/main/webapp/WEB-INF/web.xml]: @wu-sheng Here I meant

Comment (wu-sheng) [reply test/plugin/scenarios/spring-3.1.x-scenario/src/main/webapp/WEB-INF/web.xml]: Got it. Do we need to revert back? I don't want the GBK is a good example :)

Comment (kezhenxu94) [reply test/plugin/scenarios/spring-3.1.x-scenario/src/main/webapp/WEB-INF/web.xml]: > Got it. Do we need to revert back? I don't want the GBK is a good example :)

I’m ok with both situations, will revert it

**Pull request merged**